altair/utils/_transformed_data.py

```diff
@@ -0,0 +1,545 @@
+from typing import List, Optional, Tuple, Dict, Iterable, overload, Union
+
+import pandas as pd
+
+from altair import (
+    Chart,
+    FacetChart,
+    LayerChart,
+    HConcatChart,
+    VConcatChart,
+    ConcatChart,
+    data_transformers,
+)
+from altair.utils.schemapi import Undefined
+
+Scope = Tuple[int, ...]
+FacetMapping = Dict[Tuple[str, Scope], Tuple[str, Scope]]
+
+MAGIC_CHART_NAME = "_vf_mark{}"
+
+
+@overload
+def transformed_data(
+    chart: Union[Chart, FacetChart],
+    row_limit: Optional[int] = None,
+    exclude: Optional[Iterable[str]] = None,
+) -> Optional[pd.DataFrame]:
+    ...
+
+
+@overload
+def transformed_data(
+    chart: Union[LayerChart, HConcatChart, VConcatChart, ConcatChart],
+    row_limit: Optional[int] = None,
+    exclude: Optional[Iterable[str]] = None,
+) -> List[pd.DataFrame]:
+    ...
+
+
+def transformed_data(chart, row_limit=None, exclude=None):
+    """Evaluate a Chart's transforms
+
+    Evaluate the data transforms associated with a Chart and return the
+    transformed data as one or more DataFrames
+
+    Parameters
+    ----------
+    chart : Chart, FacetChart, LayerChart, HConcatChart, VConcatChart, or ConcatChart
+        Altair chart to evaluate transforms on
+    row_limit : int (optional)
+        Maximum number of rows to return for each DataFrame. None (default) for unlimited
+    exclude : iterable of str
+        Set of the names of charts to exclude
+
+    Returns
+    -------
+    pandas DataFrame or list of pandas DataFrames or None
+        If input chart is a Chart or Facet Chart, returns a pandas DataFrame of the
+        transformed data. Otherwise, returns a list of pandas DataFrames of the
+        transformed data
+    """
+    try:
+        from vegafusion import runtime, get_local_tz, get_inline_datasets_for_spec  # type: ignore
+    except ImportError as err:
+        raise ImportError(
+            "transformed_data requires the vegafusion-python-embed and vegafusion packages\n"
+            "These can be installed with pip using:\n"
+            "    pip install vegafusion[embed]\n"
+            "Or with conda using:\n"
+            "    conda install -c conda-forge vegafusion-python-embed vegafusion"
+        ) from err
+
+    if isinstance(chart, Chart):
+        # Add dummy mark if None specified to satisfy Vega-Lite
```

Comment:
Reworded comment in aabf5d6e4f7fc62fc49c4bb21b4c7a0f867e85a4.

The reason to do this is so that it's possible to call `chart.transformed_data()` on a chart with transforms even if no mark has been defined. For example:

```python
import pandas as pd
chart = alt.Chart(
    pd.DataFrame({"a": [1, 2, 3], "b": ["A", "BB", "CCC"]})
)
chart.transform_filter("datum.a > 1")._transformed_data()
```
|    |   a | b   |
|---:|----:|:----|
|  0 |   2 | BB  |
|  1 |   3 | CCC |

Not sure if this will be a common use case, but I think it's neat to be able to use an Altair chart kind of like a DataFrame this way.

Reply:
This is not a chart anymore😉! Thinking aloud, without judging: I always thought it was good that dataframe libraries make a clear distinction between dataframe operations and visualizations. With the powerful options introduced in this PR, altair is slowly turning into a dataframe library as well.

Reply:
I like this concept. I don't think this it is possible to run transforms in Vega-Lite without defining a mark. 

Naming is also an issue. `alt.Data` and `alt.Dataset` already exists. Maybe we can extend these? Worth raising a follow up issue on this!

Reply:
> I don't think this it is possible to run transforms in Vega-Lite without defining a mark

Yes, this is why I add a mark when none is defined before calling vega-lite through vl-convert.

> Naming is also an issue. alt.Data and alt.Dataset already exists

Or maybe just `alt.DataFrame`?  I'll open a follow up issue.